### ISSUE TITLE
Generic async data source for sorting, pagination and searching

### ISSUE THREAD

**Issue by salamaashoush** [grid-js/gridjs#145]
Some cases like mine we have our own API client that do data fetching/sorting/pagination from server 

A generic solution to allowing grid users to hook their own data source instead of fetching the data from gridjs code
```javascript
const grid = new Grid({
// already supported
data: () => somedatasource.getData(),
  columns: ['Pokemon', 'URL'],
  pagination: {
    limit: 5,
    server: {
      // would be useful to have similar API like this, same also for sorting and searching
      getData: (prev, page, limit) => somedatasource.getPaginatedData(prev,page,limit),
   }
  }
});
```
This is similar API from ag-grid for reference https://www.ag-grid.com/javascript-grid-server-side-model-datasource/

**Comment by afshinm** [comment]
You can use the `data` config and pass your data to the instance, see https://gridjs.io/docs/examples/import-async. You can then enable search or pagination but it will be a client-side sorting and pagination.

But I can see that you want to use server-side pagination as well which makes sense. It is possible to write a custom Storage (e.g. https://github.com/grid-js/gridjs/blob/master/src/storage/server.ts) and use it in Grid configs which will basically consume your custom API. I will add an example to the doc :+1:

**Comment by cedriking** [comment]
Hey @afshinm , I'm looking for this same option. My first call returns IDs of elements and then using GridJS I have a `data: async` that goes and fetch each ID's data. Being able to just fetch the ones that are going to be shown right away (pagination) will be amazing! Looking forward to your example 💯

**Comment by afshinm** [comment]
gotcha @cedriking. Could you please paste your code here as well? I will do my best to come up with a solution soon.

**Comment by defusevenue** [comment]
I'm currently looking for a solution to this as well. I came up with a few possible solutions:

1. Add events for pagination, sorting, and search.
2. Pass url from [pagination](https://gridjs.io/docs/examples/server-side-pagination), [sorting](https://gridjs.io/docs/examples/server-side-sort), and [search](https://gridjs.io/docs/examples/server-side-search) to the data function as parameters. Then in the data function if these variables exist we can pass it onto the server. (Recommended)

Although creating a custom storage is a viable solution, I believe the solutions I provided above would be much easier to integrate and provide backwards compatibility for future releases.
